### PR TITLE
Using DruidExceptions in MSQ (changes related to the Broker)

### PR DESCRIPTION
**Pull request by LakshSingla**
### Description

Most of the places in MSQ currently rely on the older error code, dating pre https://github.com/apache/druid/pull/14004. This is an attempt to fix the bits of the error code in the places where bits of MSQ reside in the broker. This extends upon the PR https://github.com/apache/druid/pull/14531.

1. The MSQTaskQueryMaker and the MSQTaskEngine might return a different error code for some invalid cases (which might not be inline with what the behavior originally was, but is correct now). For example, setting `maxNumTasks` to 1 would return a 500 error code, which is not right since its an invalid user input. This was the original behavior, and this PR changes that. 
2. The EXTERN function returns a 400 error code if the parameters to it are incorrect (which is in line with what its behavior originally was and is correct) 

Note: This PR doesn't unify the DruidExceptions with the MSQException

#### Release note
MSQ engine returns correct error codes for invalid user inputs in the query context

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/util/MultiStageQueryContext.java [hunk 1]
```diff
@@ -197,10 +196,7 @@ public static int getMaxNumTasks(final QueryContext queryContext)
 
   public static Object getDestination(final QueryContext queryContext)
   {
-    return queryContext.get(
-        CTX_DESTINATION,
-        DEFAULT_DESTINATION
-    );
+    return queryContext.get(CTX_DESTINATION);
```
Comment (somu-imply) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/util/MultiStageQueryContext.java:199]: Was this intentional ?

File: sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalOperatorConversion.java [hunk 2]
```diff
@@ -106,28 +106,72 @@ public ExternalTableSpec apply(
...
+            inputFormat,
             rowSignature,
             inputSource::getTypes
         );
       }
-      catch (JsonProcessingException e) {
-        throw DruidException.forPersona(DruidException.Persona.USER)
-                            .ofCategory(DruidException.Category.INVALID_INPUT)
-                            .build(e, e.getMessage());
+      catch (ArgumentAndException e) { // We can triage out the error to one of the argument passed to the EXTERN function
```
[the remark this answers is not in the document]
Comment (LakshSingla) [reply sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalOperatorConversion.java:146]: Ah yes! Will update the change

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/util/MultiStageQueryContext.java (hunk 1 above)
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/util/MultiStageQueryContext.java:199]: Yes, this is a minor refactor, the default destination is null, however, we needn't supply null to the default value in queryContext.get, as that's what queryContext.get() returns if it doesn't find the context param that we have requested.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java [hunk 3]
```diff
@@ -62,6 +61,7 @@ public class MSQTaskSqlEngine implements SqlEngine
                   .addAll(NativeSqlEngine.SYSTEM_CONTEXT_PARAMETERS)
                   .add(MultiStageQueryContext.CTX_DESTINATION)
                   .add(QueryKitUtils.CTX_TIME_COLUMN_NAME)
+                  .add(MSQTaskQueryMaker.USER_KEY)
```
[the remark this answers is not in the document]
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java:64]: Yes, this is unrelated to the change that this PR intends so I should have mentioned that 😅. We are reserving this key for usage in MSQ, therefore adding this to a reserved list.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java [hunk 4]
```diff
@@ -125,17 +125,20 @@ public boolean featureAvailable(EngineFeature feature, PlannerContext plannerCon
       case SCAN_NEEDS_SIGNATURE:
         return true;
       default:
-        throw new IAE("Unrecognized feature: %s", feature);
+        throw DruidException
+            .forPersona(DruidException.Persona.DEVELOPER)
+            .ofCategory(DruidException.Category.DEFENSIVE)
+            .build("Unrecognized feature: %s", feature);
```
Comment (imply-cheddar) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java:131]: This doesn't seem defensive either?  This message seems intended for the admin or the operator perhaps?  That or it could be end user, do we document these "features" or are they intended to be hidden?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java [hunk 5]
```diff
@@ -169,23 +172,26 @@ public QueryMaker buildQueryMakerForInsert(
...
+  private static void validateSelect(final PlannerContext plannerContext)
   {
     if (plannerContext.queryContext().containsKey(DruidSqlInsert.SQL_INSERT_SEGMENT_GRANULARITY)) {
-      throw new ValidationException(
-          StringUtils.format("Cannot use \"%s\" without INSERT", DruidSqlInsert.SQL_INSERT_SEGMENT_GRANULARITY)
-      );
+      throw DruidException
+          .forPersona(DruidException.Persona.DEVELOPER)
+          .ofCategory(DruidException.Category.DEFENSIVE)
+          .build("Cannot use \"%s\" without INSERT", DruidSqlInsert.SQL_INSERT_SEGMENT_GRANULARITY);
```
Comment (imply-cheddar) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java:186]: Once again, I think this is an `InvalidSqlException`, not defensive.  Also, the message doens't follow our conventions.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java [hunk 6]
```diff
@@ -199,12 +205,10 @@ private static void validateInsert(
...
-              StringUtils.format(
-                  "Field \"%s\" must be of type TIMESTAMP (was %s)",
-                  ColumnHolder.TIME_COLUMN_NAME,
-                  timeType
-              )
+          throw InvalidSqlInput.exception(
+              "Field \"%s\" must be of type TIMESTAMP (was %s)",
+              ColumnHolder.TIME_COLUMN_NAME,
+              timeType
           );
```
Comment (imply-cheddar) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java:212]: This message doesn't follow our conventions.  In this case, just make it

```
"Field [%s] was the wrong type [%s], expected TIMESTAMP"
```

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java [hunk 7]
```diff
@@ -220,13 +224,18 @@ private static void validateInsert(
...
-      throw new ValidationException(
-          StringUtils.format(
-              "Invalid segmentGranularity: %s",
-              plannerContext.queryContext().get(DruidSqlInsert.SQL_INSERT_SEGMENT_GRANULARITY)
-          ),
-          e
-      );
+      // This is a defensive check as the DruidSqlInsert.SQL_INSERT_SEGMENT_GRANULARITY in the query context is
+      // populated by Druid. If the user entered an incorrect granularity, that should have been flagged before reaching
+      // here
```
Comment (imply-cheddar) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java:229]: You can put all of this into the message instead of a comment.  That way, if this ever does get fired, the person who sees it will know without needing to find this comment in the code.  Also, if it's in the message, you don't need a comment in the code.


For example, you could say "Unable to parse the segmentGranularity [%s].  This should have been created by class XYZ, so this exception most likely indicates a bug over there."

Given that this is a developer-facing message, it's perfectly fine to include such details about the code and expectations of the code.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java [hunk 8]
```diff
@@ -237,11 +246,9 @@ private static void validateInsert(
...
     if (hasSegmentGranularity && timeFieldIndex < 0) {
-      throw new ValidationException(
-          StringUtils.format(
-              "INSERT queries with segment granularity other than \"all\" must have a \"%s\" field.",
-              ColumnHolder.TIME_COLUMN_NAME
-          )
+      throw InvalidInput.exception(
+          "INSERT queries with segment granularity other than \"all\" must have a \"%s\" field.",
+          ColumnHolder.TIME_COLUMN_NAME
       );
```
Comment (imply-cheddar) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java:252]: This message doesn't follow our conventions for error messages.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java [hunk 9]
```diff
@@ -150,16 +152,12 @@ public Response doPost(final SqlQuery sqlQuery, @Context final HttpServletReques
...
-                          .build(
-                              StringUtils.format(
-                                  "The statement sql api only supports sync mode[%s]. Please set context parameter [%s=%s] in the context payload",
-                                  ExecutionMode.ASYNC,
-                                  QueryContexts.CTX_EXECUTION_MODE,
-                                  ExecutionMode.ASYNC
-                              )
-                          )
+            InvalidInput.exception(
+                "The statement sql api only supports sync mode[%s]. Please set context parameter [%s=%s] in the context payload",
```
Comment (imply-cheddar) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java:156]: I'm not sure I know what the `statement sql api` is?  If it's only supposed to be one thing and this code  knows what that one thing is and the user got their request to this line of the code, why does the user need to change their query to only do the one thing that this code can do?  I.e. why not just ignore the context parameter entirely?  Or overwrite it with the one value that this code can do?

File: processing/src/main/java/org/apache/druid/error/InvalidInput.java [hunk 10]
```diff
@@ -35,7 +35,7 @@ public static DruidException exception(Throwable t, String msg, Object... args)
   private final String msg;
   private final Object[] args;
 
-  public InvalidInput(
+  protected InvalidInput(
```
Comment (imply-cheddar) [inline processing/src/main/java/org/apache/druid/error/InvalidInput.java:38]: Why this change?  Please back it out.  `protected` pretty much doesn't help anybody, might as well be public.

File: sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalOperatorConversion.java [hunk 11]
```diff
@@ -106,28 +106,72 @@ public ExternalTableSpec apply(
         }
         final RowSignature rowSignature;
         if (columns != null) {
-          rowSignature = Columns.convertSignature(columns);
+          try {
+            rowSignature = Columns.convertSignature(columns);
+          }
+          catch (IAE e) {
+            throw new ArgumentAndException("columns", e);
+          }
```
Comment (imply-cheddar) [inline sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalOperatorConversion.java:114]: Instead of this throw/catch to just throw/catch again, why not create a private method

```
public static InvalidInput badArgumentException(Throwable cause, String fieldName)
{
  return InvalidInput.exception(
            cause,
            "Invalid value for the field [%s]. Reason: [%s]",
            fieldName,
            e.getMessage()
        );
}
```

Or, perhaps better yet would be to make the place that's throwing the IAE actually throw the right DruidException instead.  You can do a `try/catch (DruidException e) { throw e.prependAndBuild("Problem converting columns") }`

**Review by imply-cheddar — CHANGES_REQUESTED** [review]

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java [hunk 12]
```diff
@@ -136,15 +136,24 @@ public QueryResponse<Object[]> runQuery(final DruidQuery druidQuery)
...
-                    + "segment graularity");
+      // This isn't populated by the user in the query context. If there was an issue with the query granularity
+      // entered by the user, it should have been flagged earlier
+      throw DruidException.forPersona(DruidException.Persona.DEVELOPER)
+                          .ofCategory(DruidException.Category.DEFENSIVE)
+                          .build(
+                              e,
+                              "Unable to deserialize the insert granularity. Please retry the query with a "
+                              + "valid segment graularity"
+                          );
```
[the remark this answers is not in the document]
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java:147]: I updated a comment to state why this is a defensive exception.

This parameter is added by Druid while processing `PARTITIONED BY _____`. User inputs whatever is in front of the partitioned by we populate the query context based on that. Therefore whatever hits this exception has been sanitized and populated by us (after already going through a transformation-serialization process). We don't ever wish to encounter this in a normal misinput situation, any exception or incorrect input from the user would have been flagged by the SQL parser or while serializing.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java (hunk 4 above)
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java:131]: These features are intended to be hidden from the user/admin/operator and are an implementation detail of Druid's planner. 
They are the properties of whatever the `SqlEngine` can or cannot support, which allows the different engines (Native and MSQ) to co-exist, and instruct the planner to plan a query in a specific way. For example, the MSQ engine doesn't support TopN queries, so by not supporting this feature, we instruct the planner to plan those as GroupBy queries.

As such, the only way we would ever encounter this is if we add a new `EngineFeature`, and we don't mention it's support in the `MSQTaskSqlEngine` or `NativeSqlEngine`, which is a developer's oversight and should have been flagged in the first place. Therefore it is a defensive exception.

Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java:131]: Adding the relevant portions from the above comment into this exception message, since these are developer-facing.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java [hunk 13]
```diff
@@ -201,15 +210,19 @@ public QueryResponse<Object[]> runQuery(final DruidQuery druidQuery)
...
 
       Granularity segmentGranularityObject;
       try {
         segmentGranularityObject = jsonMapper.readValue((String) segmentGranularity, Granularity.class);
       }
       catch (Exception e) {
-        throw new ISE("Unable to convert %s to a segment granularity", segmentGranularity);
+        throw DruidException.forPersona(DruidException.Persona.DEVELOPER)
+                            .ofCategory(DruidException.Category.DEFENSIVE)
+                            .build("Unable to convert %s to a segment granularity", segmentGranularity);
```
[the remark this answers is not in the document]
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java:225]: Good point on keeping the exception. I updated the exception message to state why this is a defensive exception.

This parameter is added by Druid while processing `PARTITIONED BY _____`. User inputs whatever is in front of the partitioned by we populate the query context based on that. Therefore whatever hits this exception has been sanitized and populated by us (after already going through a transformation-serialization process). We don't ever wish to encounter this in a normal misinput situation, any exception or incorrect input from the user would have been flagged by the SQL parser or while serializing.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java (hunk 5 above)
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java:186]: Updated the error message stating where the bug might be and why this is a defensive exception.

File: processing/src/main/java/org/apache/druid/error/InvalidInput.java (hunk 10 above)
Comment (LakshSingla) [reply processing/src/main/java/org/apache/druid/error/InvalidInput.java:38]: Changed this because we had a couple of static InvalidInput exception creators that mimic this signature `exception(...)`, therefore any piece of code that intends to call this constructor can use the creator instead for the time being.

Reverted back the change in my newer commit.

File: sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalOperatorConversion.java (hunk 11 above)
Comment (LakshSingla) [reply sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalOperatorConversion.java:114]: Seems a lot cleaner. Since the places throwing exceptions are primarily object mappers, I added the static method for creating the appropriate druid exception message.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java (hunk 9 above)
Comment (cryptoe) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java]: Since the sql statement api will support the "SYNC" mode soon, I wanted the user to set the flag explicitly. 

Rolling upgrades then become easy since we donot have a default mode/overridden mode. 

The mode also changes the way the results are fetched. 
If the mode set is async, the results are fetched via the results api. 
If the mode set is 'sync', then the results are streamed back in the same post request. (At-least, that's how I envision it. )

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java [hunk 14]
```diff
@@ -227,7 +240,9 @@ public QueryResponse<Object[]> runQuery(final DruidQuery druidQuery)
       );
     } else {
       if (ctxDestination != null && !DESTINATION_REPORT.equals(ctxDestination)) {
-        throw new IAE("Cannot SELECT with destination [%s]", ctxDestination);
+        throw DruidException.forPersona(DruidException.Persona.DEVELOPER)
+                            .ofCategory(DruidException.Category.DEFENSIVE)
+                            .build("Cannot SELECT with destination [%s]", ctxDestination);
```
[the remark this answers is not in the document]
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java:252]: Regarding this, I saw the SYSTEM_CONTEXT has the ctxDestination key, therefore I assumed that it is set by Druid, hence reaching here would have been incorrect. There were some changes merged later which I'll go through to see if this is set internally or by the user.

[the remark this answers is not in the document]
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java]: Rechecked this code, the "destination" query context is no longer valid, therefore removing these checks and the exception (cc: @cryptoe since its related to your changes)

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java (hunk 9 above)
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java]: I went through the code, and ExecutionMode is currently unused. I have left it as is since I am not aware if it will be utilized later, however, I changed the exception to defensive since this parameter isn't supported and unset in the first place.

(cc @cryptoe, I can clean this up in case we can remove the class)

**Comment by LakshSingla** [comment]
@cheddar Thanks a lot for the review, I have revised the PR with the changes, and better exception messages.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java [hunk 15]
```diff
@@ -228,16 +232,17 @@ public QueryResponse<Object[]> runQuery(final DruidQuery druidQuery)
...
       final MSQSelectDestination msqSelectDestination = MultiStageQueryContext.getSelectDestination(sqlQueryContext);
       if (msqSelectDestination.equals(MSQSelectDestination.TASK_REPORT)) {
         destination = TaskReportMSQDestination.instance();
       } else if (msqSelectDestination.equals(MSQSelectDestination.DURABLE_STORAGE)) {
         destination = DurableStorageMSQDestination.instance();
       } else {
-        throw new IAE("Cannot SELECT with destination [%s]", msqSelectDestination.name());
+        throw InvalidInput.exception(
+            "Unsupported select destination [%s] provided in the query context. MSQ can currently write the select results to "
+            + "\"taskReport\" and \"durableStorage\"",
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java:243]: You could use the enum names of `MSQSelectDestination` here

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java [hunk 16]
```diff
@@ -842,22 +826,9 @@ private static void throwIfQueryIsNotSuccessful(String queryId, TaskStatusPlus s
...
-                          .ofCategory(DruidException.Category.INVALID_INPUT)
-                          .build(
-                              StringUtils.format(
-                                  "The statement sql api only supports sync mode[%s]. Please set context parameter [%s=%s] in the context payload",
-                                  ExecutionMode.ASYNC,
-                                  QueryContexts.CTX_EXECUTION_MODE,
-                                  ExecutionMode.ASYNC
-                              )
-                          );
+      throw DruidException.defensive("[%s] is not supported. It should not be set", QueryContexts.CTX_EXECUTION_MODE);
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java:831]:
```suggestion
      throw InvalidInput.exception(
                                  "The statement sql api only supports execution mode[%s]. Please set context parameter [%s=%s] in the context payload",
                                  ExecutionMode.ASYNC,
                                  QueryContexts.CTX_EXECUTION_MODE,
                                  ExecutionMode.ASYNC
                              );
```

This should be invalid input exception cause we need to tell the user to set another execution mode.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java [hunk 17]
```diff
@@ -842,22 +826,23 @@ private static void throwIfQueryIsNotSuccessful(String queryId, TaskStatusPlus s
...
-                              StringUtils.format(
-                                  "The statement sql api only supports sync mode[%s]. Please set context parameter [%s=%s] in the context payload",
-                                  ExecutionMode.ASYNC,
-                                  QueryContexts.CTX_EXECUTION_MODE,
-                                  ExecutionMode.ASYNC
-                              )
-                          );
+      throw InvalidInput.exception(
+          "The SQL statement API currently does not support the provided execution mode [%s]. "
+          + "Please set \"%s\" to [%s] in the query context",
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java:841]: lets replace `\"%s\"` to `[%s]` everywhere

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java [hunk 18]
```diff
@@ -842,22 +826,23 @@ private static void throwIfQueryIsNotSuccessful(String queryId, TaskStatusPlus s
...
-    ExecutionMode executionMode = queryContext.getEnum(
-        QueryContexts.CTX_EXECUTION_MODE,
-        ExecutionMode.class,
-        null
-    );
+    ExecutionMode executionMode = queryContext.getEnum(QueryContexts.CTX_EXECUTION_MODE, ExecutionMode.class, null);
+
+    if (executionMode == null) {
+      throw InvalidInput.exception(
+          "Execution mode is not provided to the SQL statement API. Please set \"%s\" in the query context",
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java:833]: We should also tell the user to set 
`"Please set \"%s\" to [%s] in the query context",` in the query context so he is clear of the next step to become successful.

File: sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalOperatorConversion.java [hunk 19]
```diff
@@ -90,44 +91,73 @@ public ExternalTableSpec apply(
...
-        final String sigValue = CatalogUtils.getString(args, SIGNATURE_PARAM);
-        if (sigValue == null && columns == null) {
-          throw new IAE(
-              "EXTERN requires either a %s value or an EXTEND clause",
-              SIGNATURE_PARAM
-          );
+      final String sigValue = CatalogUtils.getString(args, SIGNATURE_PARAM);
+      if (sigValue == null && columns == null) {
+        throw InvalidInput.exception(
+            "EXTERN requires either a %s value or an EXTEND clause",
```
Comment (cryptoe) [inline sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalOperatorConversion.java:97]: [%s] square brackets

File: sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalOperatorConversion.java [hunk 20]
```diff
@@ -90,44 +91,73 @@ public ExternalTableSpec apply(
...
+            SIGNATURE_PARAM
+        );
+      }
+      if (sigValue != null && columns != null) {
+        throw InvalidInput.exception(
+            "EXTERN requires either a %s value or an EXTEND clause, but not both",
```
Comment (cryptoe) [inline sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalOperatorConversion.java:103]: [%s] square brackets

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java [hunk 21]
```diff
@@ -260,11 +258,7 @@ public Response doGetStatus(
...
       log.debug("Got forbidden request for reason [%s]", e.getErrorMessage());
-      return buildNonOkResponse(
-          DruidException.forPersona(DruidException.Persona.USER)
-                        .ofCategory(DruidException.Category.FORBIDDEN)
-                        .build(Access.DEFAULT_ERROR_MESSAGE)
-      );
+      return buildNonOkResponse(Forbidden.exception());
     }
     catch (Exception e) {
       log.warn(e, "Failed to handle query: %s", queryId);
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/resources/SqlStatementResource.java:264]:
```suggestion
      log.warn(e, "Failed to handle query: [%s]", queryId);
```

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java [hunk 22]
```diff
@@ -220,13 +223,18 @@ private static void validateInsert(
...
-          e
-      );
+      // This is a defensive check as the DruidSqlInsert.SQL_INSERT_SEGMENT_GRANULARITY in the query context is
+      // populated by Druid. If the user entered an incorrect granularity, that should have been flagged before reaching
+      // here
+      throw DruidException.forPersona(DruidException.Persona.DEVELOPER)
+                          .ofCategory(DruidException.Category.DEFENSIVE)
+                          .build(
+                              e,
+                              "Invalid %s: %s",
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskSqlEngine.java:233]:
```suggestion
                              "Invalid [%s]: [%s]",
```

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java [hunk 23]
```diff
@@ -228,16 +232,19 @@ public QueryResponse<Object[]> runQuery(final DruidQuery druidQuery)
...
       final MSQSelectDestination msqSelectDestination = MultiStageQueryContext.getSelectDestination(sqlQueryContext);
       if (msqSelectDestination.equals(MSQSelectDestination.TASK_REPORT)) {
         destination = TaskReportMSQDestination.instance();
       } else if (msqSelectDestination.equals(MSQSelectDestination.DURABLE_STORAGE)) {
         destination = DurableStorageMSQDestination.instance();
       } else {
-        throw new IAE("Cannot SELECT with destination [%s]", msqSelectDestination.name());
+        throw InvalidInput.exception(
+            "Unsupported select destination [%s] provided in the query context. MSQ can currently write the select results to "
+            + "\"%s\" and \"%s\"",
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/sql/MSQTaskQueryMaker.java:243]:
```suggestion
            + "[%s] and [%s]",
```

**Review by cryptoe — COMMENTED** [review]
Left some nit's. 
LGTM otherwise.

**Review by cryptoe — APPROVED** [review]
Changes LGTM.
